File: scripts/__tests__/delegateConsumerInventory.test.ts
import { installDelegateSandbox } from '../../packages/gateway/src/__tests__/helpers/delegateSandbox.js'
const sandbox = installDelegateSandbox()

/** Real original SQLite factory + Python kernel; Docker metadata is an explicit seam. */
import assert from 'node:assert/strict'
import { spawnSync } from 'node:child_process'
import { mkdirSync, writeFileSync } from 'node:fs'
import { join } from 'node:path'
import { test, type TestContext } from 'node:test'
import { fileURLToPath } from 'node:url'
import { DelegateDurableDb } from '../../packages/gateway/src/delegateDurable.js'
import { DelegateJobStore } from '../../packages/gateway/src/delegateJobs.js'

const adapter = fileURLToPath(new URL('./fixtures/delegateInventory.fixture.py', import.meta.url))
const kernel = fileURLToPath(new URL('../delegate-consumer-compat.py', import.meta.url))
const target = '/home/agent/.openclaude'

function fixture(t: TestContext) {
  const root = join(sandbox.root, 'volume'); mkdirSync(root)
  const master = join(sandbox.root, 'master.db')
  const volumes = [{ Name: 'oc-v5-data-u3', Mountpoint: root, Driver: 'local', Scope: 'local', Options: null }]
  const container = { Id: 'a'.repeat(64), Name: '/oc-v5-u3', Image: 'sha256:' + 'b'.repeat(64),
    Config: { User: '1000:1000', Labels: { 'com.openclaude.v3.managed': '1',
      'com.openclaude.v3.uid': '3', 'com.openclaude.runtime_channel': 'v5' }, Env: ['HOME=/home/agent'] },
    HostConfig: { RestartPolicy: { Name: 'no' } },
    Mounts: [{ Type: 'volume', Name: volumes[0]!.Name, Source: root, Destination: target, RW: true }],
    State: { Status: 'exited', Pid: 0, StartedAt: '2026-01-01T00:00:00Z' } }
  const metadata = join(sandbox.root, 'legacy.json')
  writeFileSync(metadata, JSON.stringify({ capabilities: [] }))
  function database(path: string, enabled: boolean) {
    const db = new DelegateDurableDb(path)
    const store = new DelegateJobStore({ durable: db, sm: true, failureInbox: enabled })
    assert.equal(db.minimumConsumer, enabled ? 2 : 1)
    assert.equal(db.loadAll().length, 0)
    t.after(() => store.close())
    return db
  }
  function inspect(containers: unknown[]) {
    const result = spawnSync('python3', [adapter], { encoding: 'utf8', timeout: 10_000,
      input: JSON.stringify({ volumes, containers, masters: [master] }),
      env: { PATH: process.env.PATH, HOME: sandbox.root, PYTHONDONTWRITEBYTECODE: '1' } })
    assert.equal(result.error, undefined); assert.equal(result.signal, null)
    assert.equal(result.status, 0, result.stderr); assert.equal(result.stderr, '')
    return JSON.parse(result.stdout) as { databases: { path: string; absent: boolean }[]; writers: unknown[] }
  }
  function classify(proof: ReturnType<typeof inspect>) {
    const result = spawnSync('python3', [kernel, ...proof.databases.flatMap(d => ['--database', d.path]),
      '--runtime-manifest', metadata, '--master-metadata', metadata], {
      encoding: 'utf8', timeout: 20_000,
      env: { PATH: process.env.PATH, HOME: sandbox.root, PYTHONDONTWRITEBYTECODE: '1' } })
    assert.equal(result.error, undefined); assert.equal(result.signal, null); assert.equal(result.stderr, '')
    assert.ok(!result.stdout.includes(sandbox.root), 'kernel never publishes private database paths')
    return { code: result.status, verdict: JSON.parse(result.stdout) as { required: number } }
  }
  return { root, master, container, database, inspect, classify }
}

test('offline sealed original volume refuses legacy while bootstrap-only inventory remains compatible', t => {
  const f = fixture(t)
  const path = join(f.root, 'delegate-jobs.db'), db = f.database(path, false)
  const before = f.inspect([])
  assert.equal(before.writers.length, 0)
  assert.deepEqual(before.databases.map(d => d.path).sort(), [path, f.master].sort())
  assert.equal(f.classify(before).code, 0)
  // Actual original enrollment seals this DB, not a handwritten PRAGMA/profile row.
  const enrolling = new DelegateJobStore({ durable: db, sm: true, failureInbox: true })
  assert.equal(enrolling.acceptsNewFailureSources, true)
  const after = f.inspect([])
  assert.equal(f.classify(after).code, 1)
  assert.equal(f.classify(after).verdict.required, 2)
  assert.equal(db.loadAll().length, 0, 'inventory does not create or rewrite jobs')
})

test('retained stopped container override on a distinct persistent bind participates in compatibility', t => {
  const f = fixture(t), custom = join(sandbox.root, 'custom'); mkdirSync(custom)
  f.database(join(f.root, 'delegate-jobs.db'), false)
  const db = f.database(join(custom, 'jobs.db'), true)
  f.container.Config.Env.push('OPENCLAUDE_DELEGATE_JOBS_DB=/custom/jobs.db')
  f.container.Mounts.push({ Type: 'bind', Name: '', Source: custom, Destination: '/custom', RW: true })
  const proof = f.inspect([f.container])
  assert.equal(proof.writers.length, 1)
  assert.equal(proof.databases.length, 3)
  assert.ok(proof.databases.some(d => d.path === join(custom, 'jobs.db') && !d.absent))
  assert.equal(f.classify(proof).code, 1)
  assert.equal(db.minimumConsumer, 2)
})

test('explicit master database cannot be omitted even when every container and volume is bootstrap-only', t => {
  const f = fixture(t)
  f.database(join(f.root, 'delegate-jobs.db'), false)
  const db = f.database(f.master, true)
  const proof = f.inspect([f.container])
  assert.equal(f.classify(proof).code, 1)
  assert.equal(db.minimumConsumer, 2)
})

File: scripts/__tests__/delegateConsumerInventory.test.py
"""Private filesystem + Docker metadata adapter, not a live daemon/cgroup test."""
import copy
import importlib.util
import json
from pathlib import Path
import subprocess
import tempfile
import time
import unittest
from unittest.mock import patch

SOURCE = Path(__file__).resolve().parents[1] / 'lib/delegate-consumer-inventory.py'
spec = importlib.util.spec_from_file_location('inventory', SOURCE)
inventory = importlib.util.module_from_spec(spec)
spec.loader.exec_module(inventory)


class Inventory(unittest.TestCase):
    def setUp(self):
        self.tmp = tempfile.TemporaryDirectory(prefix='oc-delegate-inventory-')
        self.addCleanup(self.tmp.cleanup)
        self.root = Path(self.tmp.name).resolve()
        self.data = self.root / 'volume'
        self.data.mkdir()
        self.master = str(self.root / 'master.db')
        self.volume = {'Name': 'oc-v5-data-u3', 'Driver': 'local', 'Scope': 'local',
                       'Options': None, 'Mountpoint': str(self.data)}
        self.container = {'Id': 'a' * 64, 'Name': '/oc-v5-u3', 'Image': 'sha256:' + 'b' * 64,
            'Config': {'Labels': {inventory.CHANNEL: 'v5', inventory.MANAGED: '1', inventory.UID: '3'},
                       'User': '1000:1000', 'Env': ['HOME=/home/agent']},
            'HostConfig': {'RestartPolicy': {'Name': 'no'}},
            'Mounts': [{'Type': 'volume', 'Name': self.volume['Name'], 'Source': str(self.data),
                        'Destination': inventory.DATA_TARGET, 'RW': True}],
            'State': {'Status': 'running', 'Pid': 123, 'StartedAt': '2026-01-01T00:00:00Z'}}

    def collect(self, volumes=None, containers=None):
        return inventory.collect(volumes if volumes is not None else [self.volume],
                                 containers if containers is not None else [self.container], [self.master])

    def test_offline_retained_volume_and_missing_master_are_included(self):
        result = self.collect(containers=[])
        self.assertEqual({d['path'] for d in result['databases']},
                         {self.master, str(self.data / 'delegate-jobs.db')})
        self.assertEqual(result['writers'], [])
        self.assertTrue(all(d['absent'] for d in result['databases']))

    def test_foreign_bind_writer_cannot_be_omitted(self):
        foreign = {'Name': '/unmanaged', 'Config': {}, 'Mounts': [
            {'Type': 'bind', 'Source': str(self.data), 'Destination': '/data', 'RW': True}]}
        with self.assertRaises(inventory.Unknown):
            self.collect(containers=[foreign])

    def test_foreign_ancestor_bind_writer_cannot_be_omitted(self):
        foreign = {'Name': '/unmanaged', 'Config': {}, 'Mounts': [
            {'Type': 'bind', 'Source': str(self.root), 'Destination': '/all', 'RW': True}]}
        with self.assertRaises(inventory.Unknown):
            self.collect(containers=[foreign])

    def test_malformed_nested_docker_data_is_typed_unknown(self):
        for field, value in [('Config', ['bad']), ('HostConfig', None), ('State', ['bad']),
                             ('Image', None), ('Mounts', [None])]:
            with self.subTest(field=field):
                bad = copy.deepcopy(self.container); bad[field] = value
                with self.assertRaises(inventory.Unknown):
                    self.collect(containers=[bad])

    def test_labels_mount_identity_and_restart_policy_fail_closed(self):
        for mutate in [lambda c: c['Config']['Labels'].update({inventory.UID: '4'}),
                       lambda c: c['HostConfig']['RestartPolicy'].update(Name='always'),
                       lambda c: c['Mounts'][0].update(Source=str(self.root)),
                       lambda c: c['Config'].update(User='root')]:
            bad = copy.deepcopy(self.container); mutate(bad)
            with self.assertRaises(inventory.Unknown): self.collect(containers=[bad])

    def test_private_override_uses_longest_exact_persistent_mount(self):
        override = self.root / 'override'; override.mkdir()
        self.container['Mounts'].append({'Type': 'bind', 'Source': str(override),
                                         'Destination': '/custom', 'RW': True})
        self.container['Config']['Env'].append('OPENCLAUDE_DELEGATE_JOBS_DB=/custom/alt.db')
        result = self.collect()
        self.assertIn(str(override / 'alt.db'), {d['path'] for d in result['databases']})
        self.assertEqual(len(result['writers']), 1)

    def test_duplicate_env_and_shadow_mounts_are_not_defaults(self):
        self.container['Config']['Env'] += ['HOME=/home/agent']
        with self.assertRaises(inventory.Unknown): self.collect()
        self.container['Config']['Env'] = ['HOME=/home/agent']
        self.container['Mounts'].append({'Type': 'tmpfs', 'Destination': inventory.DATA_TARGET + '/x'})
        self.container['Config']['Env'].append('OPENCLAUDE_DELEGATE_JOBS_DB=' + inventory.DATA_TARGET + '/x/job.db')
        with self.assertRaises(inventory.Unknown): self.collect()

    def test_symlink_database_and_swapped_root_fail_closed(self):
        db = self.data / 'delegate-jobs.db'; db.symlink_to(self.root / 'absent')
        with self.assertRaises(inventory.Unknown): self.collect()
        db.unlink(); proof = self.collect()
        self.data.rename(self.root / 'old-volume'); self.data.mkdir()
        with self.assertRaises(inventory.Unknown): inventory.revalidate(proof)

    def test_network_volume_duplicate_volume_and_missing_master_reject(self):
        bad = {**self.volume, 'Options': {'type': 'nfs'}}
        with self.assertRaises(inventory.Unknown): self.collect(volumes=[bad])
        with self.assertRaises(inventory.Unknown): self.collect(volumes=[self.volume, self.volume])
        with self.assertRaises(inventory.Unknown): inventory.collect([], [], [])

    def test_capture_checks_fixed_local_daemon_and_complete_listing(self):
        responses = ['oc-v5-data-u3\n', 'a' * 64 + '\n', json.dumps([self.volume]),
                     json.dumps([self.container]), 'oc-v5-data-u3\n', 'a' * 64 + '\n']
        calls = []
        def run(argv, **kwargs):
            calls.append((argv, kwargs))
            return subprocess.CompletedProcess(argv, 0, responses.pop(0).encode())
        with patch.object(inventory.os, 'geteuid', return_value=0), patch.object(inventory.subprocess, 'run', side_effect=run):
            result = inventory.capture_local([self.master], time.monotonic() + 5)
        self.assertEqual(len(result['writers']), 1)
        self.assertEqual(len(calls), 6)
        for argv, kwargs in calls:
            self.assertEqual(argv[:2], ['/usr/bin/docker', '--host=unix:///var/run/docker.sock'])
            self.assertNotIn('DOCKER_HOST', kwargs['env'])
            self.assertGreater(kwargs['timeout'], 0)

    def test_timeout_never_yields_partial_inventory(self):
        with patch.object(inventory.os, 'geteuid', return_value=0), patch.object(inventory.subprocess, 'run',
             side_effect=subprocess.TimeoutExpired('private-docker', 0.1)):
            with self.assertRaises(inventory.Unknown):
                inventory.capture_local([self.master], time.monotonic() + 0.1)

    def test_master_bind_alias_is_not_an_unrelated_writer(self):
        alias = self.root / 'alias'; alias.symlink_to(self.root, target_is_directory=True)
        foreign = {'Name': '/unmanaged', 'Config': {}, 'Mounts': [
            {'Type': 'bind', 'Source': str(alias / 'master.db'), 'Destination': '/db', 'RW': True}]}
        with self.assertRaises(inventory.Unknown): self.collect(containers=[foreign])

    def test_bad_inspect_or_changed_full_list_returns_no_partial_proof(self):
        for inspect, final in [('{"wrong":1}', ''), (json.dumps([self.volume]), 'new-volume\n')]:
            outputs = ['oc-v5-data-u3\n', '', inspect, final, '']
            def run(argv, **kwargs):
                return subprocess.CompletedProcess(argv, 0, outputs.pop(0).encode())
            with patch.object(inventory.os, 'geteuid', return_value=0), patch.object(inventory.subprocess, 'run', side_effect=run):
                with self.assertRaises(inventory.Unknown):
                    inventory.capture_local([self.master], time.monotonic() + 5)

    def test_unhashable_mount_and_state_fields_never_escape_as_type_errors(self):
        for mutate in [lambda c: c['Mounts'][0].update(Name=[]),
                       lambda c: c['State'].update(Status=[])]:
            c = copy.deepcopy(self.container); mutate(c)
            with self.assertRaises(inventory.Unknown): self.collect(containers=[c])


if __name__ == '__main__':
    unittest.main()
